CAMARA API Design Guide document (#441)
camaraproject/Commonalities: documentation/CAMARA-API-Design-Guide.md

The `servers` object shall have the following content:

```yaml
servers:
  # apiRoot variable and the fixed base path containing <api-name> and <api-version> as defined below  
  - url: {apiRoot}/<api-name>/<api-version>
    variables:
      apiRoot:
        default: http://localhost:9091
        description: API root, defined by the service provider, e.g. `api.example.com` or `api.example.com/somepath`
```

If more than one server object instance is listed, the `servers[*].url` property of each instance must have the same string for the `api-name` and `api-version`' in respective instance.

#### 5.5.1. api-name

`api-name` is what is specified as the base path, prior to the API version, in the `servers[*].url` property.

For the below example, the API-name would be `location-verification`.
 ```
 /location-verification/v1
```
`api-name` is a kebab-case string used to create unique names or values of objects and parameters related to given API.

#### 5.5.2. api-version

`api-version` shall be same as the `version` field in the `info` object and follow the rules defined in [7.2. API version in URL (OAS servers object)](#72-api-version-in-url-oas-servers-object).

> 

Author: It seems obvious, but it makes sense to make  #### 5.5.1. api-name and #### 5.5.2. api-version  coherent.
Updated 5.5.2.

It is recommended that the identifier have a similar morphology on all endpoints. For example, “`xxxxId`”, where xxx is the name of the entity, it references:
   - ```/users/{userId}```
   - ```/accounts/{accountId}```
   - ```/vehicles/{vehicleId}```
   - ```/users/{userId}/vehicles/{vehicleId}```

#### 5.7.2. Operations
A resource has multiple operations identified by HTTP Verbs: `GET`, `PUT`, `POST`, `PATCH`, `DELETE`.

Summary must be defined on each operation.
Functionality methods must have a description.

Allowed content type (“application/json”, “text/xml”…) should be defined. 

OperationIds are defined in lowerCamelCase: For example: `helloWorld`.

#### 5.7.3. Tags

`tags` object for each API operation is optional - Title Case is the recommended style for tag names.

#### 5.7.4. Parameters

Parameter names are defined in lowerCamelCase: For example: `SessionId`.
All parameters must have a description.
All properties within the object must have a description.

#### 5.7.5. Request bodies

'GET' and 'DELETE' HTTP methods must not accept a 'requestBody' attribute.

#### 5.7.6. Responses

> All response objects must have a description.

Author: I propose to continue in https://github.com/camaraproject/Commonalities/issues/445

### 5.8. Components

Common definitions in the global components section

#### 5.8.1. Schemas

UpperCamelCase should be used for schema names.
All properties within the object must have a description.

#### 5.8.2. Responses

> UpperCamelCase should be used for response names.

Reviewer: ```suggestion
UpperCamelCase should be used for response names.
All properties within the object must have a description.
```

#### 5.8.3. Parameters

> 

Reviewer: Maybe we can add here: Same comments as per section 5.7.4

##### Content-Type Header - clarification

The character set supported must only be UTF-8. The [JSON Data Interchange Format (RFC 8259)](https://datatracker.ietf.org/doc/html/rfc8259#section-8.1) states the following

```
JSON text exchanged between systems that are not part of a closed
ecosystem MUST be encoded using UTF-8 [RFC3629].

Previous specifications of JSON have not required the use of UTF-8
when transmitting JSON text.  However, the vast majority of JSON-
based software implementations have chosen to use the UTF-8 encoding,
to the extent that it is the only encoding that achieves
interoperability.
```

Implementers may include the UTF-8 character set in the Content-Type header value, for example, "`application/json; charset=utf-8`". However, the preferred format is to specify only the MIME type, such as "`application/json`". Regardless, the interpretation of the MIME type as UTF-8 is mandatory, even when only "`application/json`" is provided.

#### 5.8.6. Security schemes

[Security schemes](https://spec.openapis.org/oas/v3.0.3#security-scheme-object) express security in OpenAPI. 
Security can be expressed for the API as a whole or for each endpoint.

> As specified in [Use of openIdConnect for securitySchemes](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md#use-of-openidconnect-for-securityschemes), all CAMARA OpenAPI files must include the following scheme definition, with an adapted `openIdConnectUrl` in its components section. The schema definition is repeated in this document for illustration purposes, the correct format must be extracted from the link above.

Reviewer 2: Note, the OpenIdConnect security scheme does not support OAuth Client Credentials off-the-shelf. The [OpenIdConnect Discovery Procotol](https://swagger.io/docs/specification/v3_0/authentication/openid-connect-discovery/)  and [Spec](https://openid.net/specs/openid-connect-discovery-1_0.html) only supports Authorization Code and Implicit grant types.

Additional text is needed, when CAMARA recommends the support for Client Credentials and CIBA.

```yaml

> components:

Reviewer 2: This example is duplicated. Might be good to have such a template in only one place.

  securitySchemes:
    openId:
      type: openIdConnect

>       openIdConnectUrl: https://example.com/.well-known/openid-configuration

Reviewer 2: Might be good, when the API developing working group gives recommendations on the grant types.

```

The key of the security scheme is arbitrary in OAS, but convention in CAMARA is to name it `openId`.

## 6. Security

### 6.1. Good practices to secure REST APIs

> The following points can serve as a checklist to design the security mechanism of the REST APIs.

Reviewer 3: These recommendations are a blend of 'security by design' (which can be applied by the API authors during API definition) and security at the implementation phase. If we are covering both. I suggest we split them into two sections to account for what the CAMARA API authors can define, and  best practice for implementers of those API definitions.

> 1. **Simple Management**. 

Reviewer 3: This point advises 'securing only the APIs that need to be secure', but point 2 advises/mandates that 'HTTPS must always be used', which means APIs have been secured to some extent (confidentiality). So what is in scope for 'the APIs that need to be secure' vs those that do not?

Reviewer 3: I don't understand this sentence: "Any time the more complex solution is made "unnecessarily", it is also likely to leave a hole.". I agree that unnecessary complexity is to be avoided, but that is not the same as securing some APIs and not others. It is also not possible to test if this guideline has been met since there is no measure for complexity or necessity defined.

   Securing only the APIs that need to be secure. Any time the more complex solution is made "unnecessarily", it is also likely to leave a hole. 

> 2. **HTTPs must be always used**. 

Reviewer 3: ```suggestion
2. **HTTPS must always be used**. 

```

>    TLS ensures the confidentiality of the transported data and that the server's hostname matches the server's SSL certificate.

Reviewer 2: Might be good to refer to _certificate validation_, not only _hostname matiching_, e.g. [link](https://www.ssl.com/guide/ssl-best-practices/). 

```suggestion
   TLS ensures the confidentiality of the transported data and clients should do a full server SSL certificate verification, incl a hostname match.

```

>    - If HTTP 2 is used to improve performance, you can even send multiple requests over a single connection, this way you will avoid the complete overhead of TCP and SSL on later requests.

Reviewer 2: ```suggestion
   - If HTTP 2 is used to improve performance, you can even send multiple requests over a single connection, this way you will avoid the complete overhead of TCP and SSL on later requests. If HTTP 1.1 is used, usage of _TCP persistent connections_ is recommended to achieve a similar performance.

```

3. **Using hash password**. 
Passwords should never be sent in API bodies,
   but if it is necessary it must be hashed to protect the system
   (or minimize damage) even if it is compromised in some hacking attempts.
   There are many hashing algorithms that can be really effective for password security,
   for example, PBKDF2, bcrypt and scrypt algorithms.

> 4. **Information must not be exposed in the URLs**

Reviewer 2: ```suggestion
4. **Sensitive information must not be exposed in the URLs**

```

Usernames, passwords, session tokens, and API keys should not appear in the URL, as this can be captured in web server logs, making them easily exploitable. For example, this URL (```https://api.domain.com/user-management/users/{id}/someAction?apiKey=abcd123456789```) exposes the API key. Therefore, never use this kind of security.

5. **Authentication and authorization must be considered**
   CAMARA uses the authentication and authorization protocols and flows as described in the [Camara Security and Interoperability Profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md).

> 6. **Add request time flags should be considered**. 

Reviewer 2: This does not sound like a security consideration. Is there another section on general considerations?

Along with other request parameters, a request timestamp can be added as a custom HTTP header in API requests.
   The server will compare the current timestamp with the timestamp of the request
   and will only accept the request if it is within a reasonable time frame
   (1–2 minutes maybe).
   - This will prevent very basic replay attacks from people trying to hack your system without changing this timestamp.

7. **Entry params validation**

>  Validates the parameters of the request in the first step before it reaches the application logic.

Reviewer 2: ```suggestion
 Validates the parameters of the request against access token information in the first step before it reaches the application logic.

```

   Put strong validation checks and reject the request immediately if validation fails.
   In the API response,
   send relevant error messages and an example of the correct input format to improve the user experience.

### 6.2. Security definition

In general, all APIs must be secured to ensure who has access to what and for what purpose.

> CAMARA uses OIDC and CIBA for authentication and consent collection and to determine whether the user has,

Reviewer 2: ```suggestion
CAMARA uses OIDC, CIBA and Client Credentials for authentication and consent collection and to determine whether the user has,

```

> CAMARA uses OIDC and CIBA for authentication and consent collection and to determine whether the user has,

Reviewer 2: ```suggestion
CAMARA uses 
- OIDC and CIBA for authentication and consent collection and to determine whether the user has, e.g. opted out of some API access,
- Client Credentials for authentication for other cases. 

```

> e.g. opted out of some API access.

Reviewer 2: ```suggestion

```
(appended to the previous line)

The [Camara Security and Interoperability Profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md#purpose) defines that a single purpose is encoded in the list of scope values. The purpose values are defined by W3C Data Privacy Vocabulary as indicated in the [Profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md#purpose-as-a-scope).

### 6.3. Expressing Security Requirements

Security requirements of an API are expressed in OpenAPI through [Security Requirement Objects](https://spec.openapis.org/oas/v3.0.3#security-requirement-object).

The following is an example of how to use the `openId` security scheme defined above as described in [Use of security property](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-API-access-and-user-consent.md#use-of-security-property):

```yaml
paths:
  {path}:
    {method}:
      ...
      security:
        - openId:
            - {scope}
```

> 

Reviewer 2: hmm, this is more a repetition. CAMARA requires the usage of the openIdConnection security scheme in other places. So, what is the additional requirement (see heading)?

| qod                   | /qos-profiles | GET    | `qod:qos-profiles:read`                           |

#### 6.6.2. API-level scopes (sometimes referred to as wildcard scopes in CAMARA)

The decision on the API-level scopes was made within the [Identity and Consent Management Working Group](https://github.com/camaraproject/IdentityAndConsentManagement) and is documented in the design guidelines to ensure the completeness of this document. 
The scopes will always be those defined in the API Specs YAML files. Thus, a scope would only provide access to all endpoints and resources of an API if it is explicitly defined in the API Spec YAML file and agreed in the corresponding API subproject. 

### 6.7. Resource access restriction

In some CAMARA APIs there are functions to create resource (via POST) and then later query them via id and/or list (with GET) or delete them (via DELETE). For example we have sessions, payments, subscriptions, etc..

For the GET and DELETE operations we must restrict the resource(s) targeted depending on the information provided in the request. Basically we consider 2 filters:

> * API client (aka ClientId)

Reviewer 2: Consistency: Would be good to use either _API client_  or _API consumer_, but not both. 

Personally, _API consumer_  or _API invoker_ clearer.

* access token

| Operation |  3-legged access token is used | 2-legged access token is used |
|-----------|--------------------------------|-------------------------------|

> | GET/{id} | - The resource queried must have been created for the end user associated with the access token. <br> - The resource queried must have been created by the same API client given in the access token. | - The resource queried must have been created by the same API client given in the access token. |

Reviewer 2: > The resource queried must have been created by the same API client given in the access token

What is the motivation for limiting it to the same API Client (ClientId)? There are role-based authorization use-cases, where elevated privileges are required for creating a resource (e.g. creating a dedicated network / a network slice) than for reading the resource?

I would believe that one enterprise (as API invoker) has multiple ClientIds, and the different authorization levels are determent when obtaining the access token...

CAMARA use the OpenAPI Specification (OAS) to describe their APIs. The following guidelines outline the restrictions and conventions to be followed within the OAS files by all CAMARA APIs.

### 1.1. Conventions
The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### 1.2. Common Vocabulary and Acronyms

| **Term**       | Description                                                                                                                                                                                                                                                                                                                                         |
|----------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------| 

> | **API**        | Application Programming Interface. It is a rule & specification group (code) that applications follow to communicate between them, used as interface among programs developed with different technologies.                                                                                                                                          |

Reviewer 4: It could be more clear with following - 'It is a set of rules and specifications that enable communication between software applications.' against 'It is a rule & specification group (code) that applications follow to communicate between them'

- Name of the data object, used to reference it in other sections.
- Data type (String, Integer, Object…).
- If the data type is string it is recommended to use appropriate modifier property `format` and/or `pattern` whenever possible. The [OpenAPI Initiative Formats Registry](https://spec.openapis.org/registry/format/) contains the list of formats used in OpenAPI specifications.
  - If the format of a string is `date-time`, the following sentence must be present in the description: `It must follow [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6) and must have time zone. Recommended format is yyyy-MM-dd'T'HH:mm:ss.SSSZ (i.e. which allows 2023-07-03T14:27:08.312+02:00 or 2023-07-03T12:27:08.312Z)`
- If the data type is an object, list of required properties.
- List of properties within the object data, including:
   - Property name
   - Property description
   - Property type (String, Integer, Object …)
   - Other properties by type:
      - String ones: min and max length
      - Integer ones: Format (int32, int64…), min value.

> In this part, the error response structure must also be defined following the guidelines in [Chapter 6. Error Responses](#6-error-responses).

Reviewer 4: Reference link is incorrect - Error response is under section 3

- `Content-Last-Key`: it allows specifying the key of the last resort provided in the response
- `X-Total-Count`: it allows indicating the total number of elements in the collection

If the server cannot meet any of the required parameters, it should return an error message.

The HTTP codes that the server will use as a response are:
- `200`: the response includes the complete list of resources
- `206`: the response does not include the complete list of resources
- `400`: request outside the range of the resource list

Petitions examples:
- `page=0 perPage=20`, which returns the first 20 resources
- `page=10 perPage=20`, which returns 20 resources from the 10th page (in terms of absolute index, 10 pages and 20 elements per page, means it will start on the 200th position as 10x20=200)

### 4.2. Sorting

Sorting the result of a query on a resource collection requires two main parameters:
Note: Services must accept and use these parameters when sorting is supported.  If a parameter is not supported, the service should return an error message.
- `orderBy`: it contains the names of the attributes on which the sort is performed, with comma separated if there is more than one criteria.

> - `order`: by default, sorting is done in descending order. 

Reviewer 4: I believe , descending order is unusual as a default (ascending is more common).

Reviewer: Maybe we can add some clarifications here (even discuss in separate issue). 
* When indicating this definition probably it was mainly thinking in `dates` or `numerical` concepts for `orderBy` where it makes sense to have descending order as default (most recent items first, higher values first).

* However, when the concept is a "string" (alphanumeric value), it makes sense to be ascending order (from "a" to "z", from "0" to "9").

For combination scenarios (sorting by more than one concept), could be several ways. I think two options would be reasonable from beginning:
- Delegate to the API WG and document in the API
- Model order "default" value based on the nature of the first concept indicated in `orderBy`

Author: @PedroDiez Thank you for proposed clarifications.

I prefer to discuss this change in separate issue.
Are any other APIs than Carrier Billing use sorting feature now?

Reviewer: I will open the issue.

So far, I have no identified any other API rather than Carrier Billing usign this feature

      openIdConnectUrl: https://example.com/.well-known/openid-configuration
```

The key of the security scheme is arbitrary in OAS, but convention in CAMARA is to name it `openId`.

## 6. Security

### 6.1. Good practices to secure REST APIs

The following points can serve as a checklist to design the security mechanism of the REST APIs.

1. **Simple Management**. 
   Securing only the APIs that need to be secure. Any time the more complex solution is made "unnecessarily", it is also likely to leave a hole. 
2. **HTTPS must always be used**. 

   TLS ensures the confidentiality of the transported data and that the server's hostname matches the server's SSL certificate.
   - If HTTP 2 is used to improve performance, you can even send multiple requests over a single connection, this way you will avoid the complete overhead of TCP and SSL on later requests.

3. **Using hash password**. 
Passwords should never be sent in API bodies,
   but if it is necessary it must be hashed to protect the system
   (or minimize damage) even if it is compromised in some hacking attempts.
   There are many hashing algorithms that can be really effective for password security,

>    for example, PBKDF2, bcrypt and scrypt algorithms.

Reviewer 4: we can refer 'Argon2' that is considered the most modern and secure instead of 'PBKDF2' which is older now.

- Do not change the field’s semantics.
- Do not change the field’s order.
- Do not change the validation rules of the request fields to more restrictive ones.
- If you use collections that can be returned with no content, then answer with an empty collection and not null.
- Layout pagination support from the start.

Make the information available:
- provide access to the new API version definition file (via a link or dedicated endpoint)
- if possible, do the same to get the currently implemented API version definition file

**As API consumer**:
- Tolerant reader: if it does not recognize a field when faced with a response from a service, do not process it, but record it through the log (or resend it if applicable).
- Ignore fields with null values.
- Variable order rule: DO NOT rely on the order in which data appears in responses from the JSON service, unless the service explicitly specifies it.
- Clients MUST NOT transmit personally identifiable information (PII) parameters in the URL. If necessary, use headers.

> ## 8. External Documentation

Reviewer 4: This section is currently empty, and I’m not sure whether Section 5.4 is referring to the same externalDocs.

Author: Yes , it is empty placeholder for other documents to be referenced in guidelines if needed.
If not needed it can be removed.
5.4 is related to OpenAPI section externalDocs for specific API.

    #...
components:
  parameters: 
    StartCreationDate:   #<-- component name
      in: query
      name: creationDate.gte    <-- query attribute for "greater - equal to" comparison 
      required: false
      schema:
        format: date-time
        type: string
    AfterCreationDate:
      in: query
      name: creationDate.gt
      required: false
      schema:
        format: date-time
        type: string
    EndCreationDate:
      in: query
      name: creationDate.lte
      required: false
      schema:
        format: date-time
        type: string
    BeforeCreationDate:
      in: query
      name: creationDate.lt
      required: false
      schema:
        format: date-time
        type: string
```

Then the parameters can be included in the query:
  - `GET /users?creationDate.gte=2021-01-01T00:00:00`
    - Find users with creationDate starting from 2021
  - `GET /users?creationDate.lt=2022-01-01T00:00:00`
    - Find users with creationDate before 2022
  - `GET /users?creationDate.gte=2020-01-01T00:00:00&creationDate.lte=2021-12-31T23:59:59`
    - Search for users created between 2020 and 2021

## 5. OpenAPI Sections

### 5.1. Reserved words

> To avoid issues with implementation using Open API generators reserved words must not be used in the following parts of an API specification:

Reviewer 2: ```suggestion
To avoid issues with implementation using Open API generators the following words must not be used in the following parts of an API specification (reserved words):

```

- Path and operation names
- Path or query parameter names
- Request and response body property names
- Security schemes
- Component names
- OperationIds

A reserved word is one whose usage is reserved by any of the following Open API generators:
- [Python Flask](https://openapi-generator.tech/docs/generators/python-flask/#reserved-words)
- [OpenAPI Generator (Java)](https://openapi-generator.tech/docs/generators/java/#reserved-words)
- [OpenAPI Generator (Go)](https://openapi-generator.tech/docs/generators/go/#reserved-words)
- [OpenAPI Generator (Kotlin)](https://openapi-generator.tech/docs/generators/kotlin/#reserved-words)
- [OpenAPI Generator (Swift5)](https://openapi-generator.tech/docs/generators/swift5#reserved-words)

### 5.2. OpenAPI Version
The API functionalities must be implemented following the specifications of the [Open API version 3.0.3](https://spec.openapis.org/oas/v3.0.3) using `api-name` as the filename and the `.yaml` or `.json` file extension.

### 5.3. Info Object

> The `info` object shall have the following content:

Reviewer 2: ```suggestion
The `info` object shall be present in each API specification and shall contain the below listed properties. 

Example `info` object:

```

```yaml
info:
  # title without "API" in it, e.g. "Number Verification"
  title: Number Verification
  # description explaining the API, part of the API documentation 
  # text explaining how to fill the "Authorization and authentication" - see section 11.6
  description: |
    This API allows to verify that the provided mobile phone number is the one used in the device. It
    verifies that the user is using a device with the same mobile phone number as it is declared.
    ### Authorization and authentication
    CAMARA guidelines defines a set of authorization flows ...
  # API version - Aligned to SemVer 2.0 according to CAMARA versioning guidelines
  version: 1.0.1
  # Name of the license and a URL to the license description
  license:
    name: Apache 2.0
    url: https://www.apache.org/licenses/LICENSE-2.0.html
  # CAMARA Commonalities minor version - x.y
  x-camara-commonalities: 0.5
```

#### 5.3.1. Title

> Title describes the API shortly. The title shall not include the term "API" in it.

Reviewer 2: ```suggestion
`info.title`  describes the API shortly. The title shall not include the term "API" in it.

```

#### 5.3.2. Description

> No special restrictions specified in CAMARA.

Reviewer 2: ```suggestion
`info.description`:  There are no special restrictions specified in CAMARA. Some section are required, e.g. on authorization.
```

#### 5.3.3. Version

> APIs shall use the [versioning-format](https://lf-camaraproject.atlassian.net/wiki/x/3yLe) as specified by the release management working group.

Reviewer 2: ```suggestion
`info.version`: APIs shall use the [versioning-format](https://lf-camaraproject.atlassian.net/wiki/x/3yLe) as specified by the release management working group.

```

#### 5.3.4. Terms of service
`termsOfService` shall not be included. API providers may add this content when documenting their APIs.

#### 5.3.5. Contact information
`contact` information shall not be included. API providers may add this content when documenting their APIs.

#### 5.3.6. License
The license object shall include the following fields:
```
license
  name: Apache 2.0
  url: https://www.apache.org/licenses/LICENSE-2.0.html
```

#### 5.3.7. Extension field

> The API shall specify the Commonalities minor release number they are compliant to, by including the `x-camara-commonalities` extension field.

Reviewer 2: ```suggestion
`info.x-camara-commonalities`: The API shall specify the Commonalities minor release number they are compliant to, by including the `x-camara-commonalities` extension field.

```

### 5.4. ExternalDocs object

> The `externalDocs` object shall have the following content:

Reviewer 2: ```suggestion
The `info.externalDocs` object shall have the following content:

```
